jbmc/src/java_bytecode/select_pointer_type.cpp

```diff
@@ -25,61 +25,60 @@ pointer_typet select_pointer_typet::convert_pointer_type(
   // a generic parameter, specialize it with concrete types
   if(!generic_parameter_specialization_map.empty())
   {
-    generic_parameter_recursion_trackingt visited;
-    const auto &type = specialize_generics(
-      pointer_type, generic_parameter_specialization_map, visited);
-    INVARIANT(visited.empty(), "recursion stack must be empty here");
-    return type;
+    return specialize_generics(
+      pointer_type, generic_parameter_specialization_map);
   }
   else
   {
     return pointer_type;
   }
 }
 
+/// Return the first concrete type instantiation if any such exists.
+///
+/// Example:
+/// `class MyGeneric<T,U> { MyGeneric<U,T> gen; T t;}`
+/// When instantiating `MyGeneric<Integer,String> my` we need to for example
+/// resolve the type of `my.gen.t`. The map would in this context contain
+/// - T -> (Integer, U)
+/// - U -> (String, T)
+///
+/// Note that the top of the stacks for T and U create a recursion T -> U,
+/// U-> T. We want to break it and specialize `T my.gen.t` to `String my.gen.t`.
+/// \param pointer_type: The pointer type to specialize
+/// \param generic_parameter_specialization_map: Map of type names to
+///   specialization stack
+/// \return The first instantiated type for the generic type or nothing if no
+///   such instantiation exists.
 pointer_typet select_pointer_typet::specialize_generics(
   const pointer_typet &pointer_type,
   const generic_parameter_specialization_mapt
-    &generic_parameter_specialization_map,
-  generic_parameter_recursion_trackingt &visited_nodes) const
+    &generic_parameter_specialization_map) const
 {
   auto parameter = type_try_dynamic_cast<java_generic_parametert>(pointer_type);
   if(parameter != nullptr)
   {
-    const irep_idt &parameter_name = parameter->get_name();
+    irep_idt parameter_name = parameter->get_name();
 
-    // avoid infinite recursion by looking at each generic argument from
-    // previous assignments
-    if(visited_nodes.find(parameter_name) != visited_nodes.end())
+    generic_parameter_specialization_mapt spec_map_copy =
+      generic_parameter_specialization_map;
```

Comment:
:bulb: You're making a copy of the map over and over again, changing a single element in it in each call. I don't have any good idea how to make this less costly though.. not so nice ideas being keeping two functions, make the code uglier at the call site, or add extra arguments to the function (depth of each stack to look at). Can you think of a better way?